Notebook read sample tags different from the tags used in the test and add comments to rego rules

revit13 opened the pull request:
/Closes #2020 

This PR makes the notebook read sample tags the same as the tags used in the notebook read test
It also add comments to rego rules.

roytman commented on manager/testdata/notebook/read-flow/pii-policy-cm.yaml:
```diff
@@ -3,17 +3,12 @@ data:
-        column_names := [input.resource.metadata.columns[i].name | input.resource.metadata.columns[i].tags.PII]
+        input.resource.metadata.tags["Purpose.finance"]
+        column_names := [input.resource.metadata.columns[i].name | input.resource.metadata.columns[i].tags["PII.Sensitive"]]
+        count(column_names) > 0
```
I compare these roles with the roles in the write scenario, both Web site and the test.
The write policy looks like :

```rego
rule[{}] {
        description := "allow read datasets"
        input.action.actionType == "read"
    }

    rule[{"action": {"name":"RedactAction","columns": column_names}, "policy": description}] {
        description := "Redact written columns tagged as sensitive in datasets tagged with finance = true. The data should not be stored in `neverland` storage account"
        input.action.actionType == "write"
        input.resource.metadata.tags.finance
        input.action.destination != "neverland"
        column_names := [input.resource.metadata.columns[i].name | input.resource.metadata.columns[i].tags.sensitive]
    }
```

Maybe we have to sync all examples and tests.

revit13 replied:
yes, as noted in https://github.com/fybrik/fybrik/pull/1994 the notebook write test still differs from the one in the site due to:

- the default policy is deny instead of allow in the site
- adminCRsNamespace is set to default namespace instead of fybrik-system in the site.

this was done to reduce the number of tests.

revit13 commented:
@roytman I added comments to make rego rules more readable. if the rule appears in the site and tests then I added the comment to the site only for now. after review I will copy the comments to the tests. Thanks

shlomitk1 commented on charts/fybrik/notebook-test-readflow.tls.values.yaml:
```diff
@@ -90,22 +90,16 @@ opaServer:
+           input.action.actionType == "read"
+           input.resource.metadata.tags["Purpose.finance"]
+           column_names := [input.resource.metadata.columns[i].name | input.resource.metadata.columns[i].tags["PII.Sensitive"]]
+           count(column_names) > 0
```
> Why do we check `count(column_names) > 0`? Or this test is with allow-by-default?

The redaction rule should not be triggered if there is no data to redact. If no rules are triggered, the default policy will be applied - allow or not.

revit13 replied:
if the default policy is `deny by default` then the `count(column_names) > 0` condition is needed to deny assets with "Purpose.finance" tag and that do not contain "PII.Sensitive" tags. because otherwise an action object with empty columns is returned while with `count(column_names) > 0` the rule will not be set and if no other rule exists then the request will be denied.
This is a link to opa playground showing this case https://play.openpolicyagent.org/p/MUAzu72S2n
When the default policy is `allow by default` the `count(column_names) > 0` condition is redundant because with this condition and without it the request will not be denied: with `count(column_names) > 0` allow by default will be applied given that no other rule exists and without `count(column_names) > 0` action is returned with empty columns.

shlomitk1 replied:
This condition is not redundant. Suppose you have no sensitive data and you decide not to deploy a redaction module. What will happen is that a Redact action will be returned (with empty columns), Fybrik will fail to find a module that does redaction, thus, failing to create a data plane. The request will be stuck in an error state waiting for a module to be deployed...

revit13 replied:
@shlomitk1 OK thanks! I will add this explanation to the condition comment per @tomersolomon1 request.

tomersolomon1 replied:
I suggest adding the following comment:
```
If column_names is empty this rule will not be triggered (and the default rule will be triggered instead). Note however that currently in the test there are columns tagged as sensitive, and furthermore the default is set to allow - so practically this behaves the same with and without this condition. However, we check the length of column_names as an example of good policy writing. Since currently even when column_names is empty a reduction module will be deployed, which will do nothing, and is therefore a waste of resources.
```

revit13 replied:
@tomersolomon1 Thanks.  I ended up adding a short explanation in a comment.

flora177 commented on charts/fybrik/files/opa-server/policy-lib/internals/lib.rego:
```diff
@@ -1,12 +1,51 @@
+#   all the conditions between the curly braces are true:
+#   Assigment IF { CONDITIONS }
+#
+# - Rule order is irrelavant.
```
test comment
